Utilities: Locate and square/circle intersection area

baperry2 opened the pull request:
These functions are handy to have available for prob source files without needing to reimplement each time:

- `locate`: find the largest index that is less than an input value in a sorted array (used for interpolation). We have this implemented a few times across PeleC, PeleLMeX, and PelePhysics. I will collapse this to a single implementation for simplicity.
- `rectangle_circle_intersection_area`: useful for problems with round uniform-velocity inlets, by computing the intersection area with a cell we can scale the inlet velocity accordingly rather than a binary based on whether the cell center is within the inlet. This allows a mss flow to be specified and kept constant regardless of the resolution with which the inlet is resolved (even if it is smaller than a single cell!)

baperry2 committed:
add some useful utilities for Prob files

baperry2 committed:
for square/circle intersection, ignore when square is far away

baperry2 committed:
revert change to locate function, keep classic behavior

baperry2 committed:
make locate host_device instead of just device

baperry2 committed:
add utility unit tests to CI

baperry2 committed:
add minimal documentation on new utilities

baperry2 committed:
Merge branch 'development' into utilities

d-montgomery commented on Testing/Exec/UtilityUnitTest/main.cpp:
```diff
@@ -0,0 +1,61 @@
+
+  // Test Unit Conversions
+  namespace m2c = pele::physics::utilities::mks2cgs;
+  namespace c2m = pele::physics::utilities::cgs2mks;
```
I think these aliases need to be moved above `main()`

baperry2 committed:
use standard build system for utility unit tests due to CUDA issue

baperry2 commented:
For the utility unit tests that I added here, I had tried to set up a simpler GNUmakefile that would only build the relevant parts of the code, but for some reason I couldn't figure out, this was causing issues with CUDA. So I've reverted it to just use the standard `GNUmakefile` and `Make.PelePhysics`.

@marchdf and @jrood-nrel  - thoughts on the approach here for unit testing utilities? It's definitely not the prettiest but maybe it's good for now?

marchdf commented:
I am fine with the slightly overly complicated gnumakefile for the unit tests. There's probably some needed amrex options that need to be set that aren't allowing you to remove as much as you would like...

baperry2 committed:
move namespace aliases outside of main function

baperry2 committed:
put some stuff back into GNUMakefile

baperry2 committed:
add nprocs arg to utilities build in ci
